### DISCUSSION TITLE
v3.0.0

### DISCUSSION THREAD

**Discussion by Daveiano** [Daveiano/weewx-wdc#121] [Announcements]
**Requires WeeWX >=v4.9.0**

- Bugfix: Forecast error using OWM Forecast [GH-100]
- Bugfix: Forecast NWS Bug: Reason: '>' not supported between instances of 'NoneType' and 'int' [GH-118]
- Bugfix: Why are the Weather intervals (week, month, year) all off by one day? [GH-116]
- Show weather data for a specific day [GH-67]
- Make diagrams configurable per context (per period) / Refactor diagrams configuration [GH-73]. Diagrams can now be configured per context, eg. day, week or month. Per default, the day and week pages do not include the `outTemp` min/max/avg diagram anymore.
- Enhancements for weewx-forecast support [GH-95]
- Make show_min, show_max, show_sum configurable for stat tiles [GH-94]
- Support for custom `data_bindings` [GH-71]
- Configurable Icons [GH-87]
- Configurable unit for diagrams [GH-88]
- Configurable Rounding [GH-89]
- Make markers configurable [GH-106]
- Automatic Refresh [GH-105]
- Support for combining different diagram types (eg. line and bar) [GH-75]. This adds an own implementation of chart rendering based directly on [D3.js](https://d3js.org/), please see https://github.com/Daveiano/weewx-wdc/wiki/Configuration#ENABLE_D3_DIAGRAMS (a Feature Flag was added to skin.conf for activating the new charts)
- Along [GH-75]: Added new climatogram (combined chart consisting of rain as bar and outTemp as line) for statistics and yearly statistics pages
- Update radar_img to allow for raw HTML (for using eg. iFrames as radar maps) [GH-108]
- Wind direction aggregation - use vecdir instead of avg [GH-119]

**Please have a look at the [wiki](https://github.com/Daveiano/weewx-wdc/wiki) for information on how to configure the new features.**

## Changes made to skin.conf since 2.3.3

This update contains like a rewrite of the skin.conf file, a complete diff can be found here (please click on "Load diff"):

https://github.com/Daveiano/weewx-wdc/compare/v2.3.3...7da0c421941c8e61e5aaf2b6d236ae7cf231009b#diff-ba225fb627dcbf577d0d1fde0f18d93e8e2cf2097c37086cd28ef23e8e7bd820

## What's Changed
* 67 show weather data for a specific day by @Daveiano in https://github.com/Daveiano/weewx-wdc/pull/91
* Added first implementation for diagrams per context. Removed temp min… by @Daveiano in https://github.com/Daveiano/weewx-wdc/pull/93
* 95 Enhancements for weewx-forecast support by @Daveiano in https://github.com/Daveiano/weewx-wdc/pull/98
* 71 specify data binding on a per tile basis by @Daveiano in https://github.com/Daveiano/weewx-wdc/pull/101
* 94 make show min show max show count etc configurable for stat tiles by @Daveiano in https://github.com/Daveiano/weewx-wdc/pull/102
* Configurable Icons by @Daveiano in https://github.com/Daveiano/weewx-wdc/pull/103
* Configurable rounding unit by @Daveiano in https://github.com/Daveiano/weewx-wdc/pull/104
* Bump http-cache-semantics from 4.1.0 to 4.1.1 by @dependabot in https://github.com/Daveiano/weewx-wdc/pull/112
* Forecast NWS Bug: Reason: '>' not supported between instances of 'NoneType' and 'int' by @Daveiano in https://github.com/Daveiano/weewx-wdc/pull/118
* 75 support for combining different diagram types eg line and bar by @Daveiano in https://github.com/Daveiano/weewx-wdc/pull/109
* windDir aggregation changed from avg to vecdir by @Jterrettaz in https://github.com/Daveiano/weewx-wdc/pull/120

## New Contributors
* @Jterrettaz made their first contribution in https://github.com/Daveiano/weewx-wdc/pull/120

**Full Changelog**: https://github.com/Daveiano/weewx-wdc/compare/v2.3.3...v3.0.0

<hr /><em>This discussion was created from the release <a href='https://github.com/Daveiano/weewx-wdc/releases/tag/v3.0.0'>v3.0.0</a>.</em>

**Comment by JonFear** [comment]
Thank you so much for the release. As expected I have made an error somewhere in the re-install! I only get a very basic page when I load my data. I have attached a picture as an example. The data is briefly shown but that is only down the LH side of the screen and then is covered by what you see in the image.

I am sure it is a school boy error but where?

Thanks

Jon
![weewx-wdc-1](https://user-images.githubusercontent.com/7197757/224479933-da80a568-f082-459b-b18d-d136437d369e.png)

**Reply by Daveiano** [reply]
Here you go, another user had the same "school boy error" (:D) yesterday: https://github.com/Daveiano/weewx-wdc/issues/126#issuecomment-1464131941

**Reply by JonFear** [reply]
Doh! Thank you! Looks nice, just fighting the forecast now...

**Reply by JonFear** [reply]
OK so the template has been uploaded to my cloud site and is available on www.wslwx.co.uk. Still needs some tweeking but looks great, thank you for your efforts.

Looks like the Stats page needs some input, acting very strange...

One thing noted by one of my users is the webcam image that I had on my previous site. I have taken a look through but cannot see anything about adding either an image or a live feed to the main page.

Is there any plans for this?

Again, my profound admiration for your work, thank you.

Jon

**Reply by Daveiano** [reply]
Thank you very much for these kind words!

> Looks like the Stats page needs some input, acting very strange...

You are looking at the Climatogram, which looks kind of strange? Yes, I noticed this also for short periods of time (< ~2 years), I will have a look at this.

> One thing noted by one of my users is the webcam image that I had on my previous site. I have taken a look through but cannot see anything about adding either an image or a live feed to the main page.

There is currently only one slot for an external image (the `radar_html`), so you can either have a radar image/video OR a webcam image, that's correct. I will put this on the list for the next feature update, sounds like a useful addition which would also benefit other users.

**Reply by JonFear** [reply]
Hi

Understood about the webcam. I guess I could add it for now as an iframe on another page which at least gives access. It may be an idea to have a web cam page as well as the cloudcam which looks directly up from my location, I now have a Sony Starvis camera facing north for any future aurora, not that we see many at 51.3N but... we are heading for maxima in the next couple of years...

If you need any data from me on the Stats side of things let me know...

BW

Jon

**Reply by Daveiano** [reply]
An extra webcam page sounds good.

Perhaps I could provide the following with the next release:

- An additional optional "Web Cam/Externals" Page which works like the current about page, the user needs to enable and edit the page
- Some basic Code Snippets in the wiki for users to easily create 2-Column layouts, 3-Column layouts, Headings or whatever

What do you think about that?

Dave

**Reply by JonFear** [reply]
Hi Dave

Yes, an optional webcam page would be brilliant and I would hope something that given your skills, quite easy to implement. 

I am currently thinking about having either a current snapshot or live video and a 24 hour time-lapse per camera. I only have the 2 camera's which I suspect is the maximum that anyone would need!

BW

Jon

**Reply by Daveiano** [reply]
Ok, so you mean in total that would be four videos/images? Two per camera?

**Reply by JonFear** [reply]
Dave, 

Yes, a snap shot which will refresh each time the page refreshes and a time lapse video created once a day for each camera.

Just working on the mechanics of getting the images and video created. Wish I still had my old fileserver, I could just run up a new VM!

Time to install Virtual Box on my W10 gateway machine...

Jon

**Reply by JonFear** [reply]
Dave, additional information. The snapshots are jpg, the timelapse files are mp4. I have created some linux bash scripts to create the files from the two cameras. I have set up the crontab file to run tonight with the snapshot files I have created during the day, we will see whether it all works tomorrow morning!
J

**Reply by Daveiano** [reply]
Hi Jon, how is it going so far with your webcam page?

**Reply by JonFear** [reply]
Hi Dave,

Not much further than creating the files. I have a snap generated every 30 seconds for each camera and this is placed in /var/www/html/weewx/snap. A copy of it is also renamed into a time stamped file for the video creation in a folder on my NAS.

Every 24 hours, there is a time lapse mp4 file created for each camera, this is placed the same snap directory from my NAS. 

Today is the first time I have the full 24 hour video's available. From first glance, one video is much bigger than the other by quite some margin (62M v 284M) so this needs to be looked at and resolved. Once that is solved it then take a look at the code on the about page and see if I can figure out what is going on!

Update: Figured out what went wrong on the larger video. One additional letter in the code can have a massive effect. Let's see what is created tonight...

Happy days...

Jon

**Reply by Daveiano** [reply]
Hi Jon,

I got some time and put together some things: 

- A switch on top of the current radar tile to switch between radar and webcams:

![Webcam Switch](https://public-images-social.s3.eu-west-1.amazonaws.com/weewx-wdc-webcam-tabs.png)

And an optional Externals/Webcams page with up to 4 slots (please ignore the example images): https://github.com/Daveiano/weewx-wdc/blob/127-add-slot-for-webcam-imageadd-optional-extra-webcam-page/test/e2e-tests/basic/radar-externals.test.ts-snapshots/Radar-and-Externals-embeds-Externals-page-1-chromium-linux.png

Everything is configurable via skin.conf. This should satisfy your needs? :)

**Reply by JonFear** [reply]
Hi Dave

That looks really nice, way above what I expected, thank you! Looked rather damp in the radar...

Is there a simple way of installing what you have generated or should I wait until the release?

As an aside, I seem to have picked up an additional "Forecast &" in the Forecast and Radar title bar. I cannot see where from. It is there whether I put the title in the skin or weewx.conf files... 

Thanks again

Jon

**Reply by Daveiano** [reply]
> Is there a simple way of installing what you have generated or should I wait until the release?

Please be patient and wait until the next release, it will arrive before April ;)

> As an aside, I seem to have picked up an additional "Forecast &" in the Forecast and Radar title bar. I cannot see where from. It is there whether I put the title in the skin or weewx.conf files...

Did you accidentally set this in skin.conf under `[Extras]radar_heading`?

**Reply by JonFear** [reply]
> > Is there a simple way of installing what you have generated or should I wait until the release?
> 
> Please be patient and wait until the next release, it will arrive before April ;)

Oh OK Then! I will sit on my hands until you say so! TBF I have to change one of the camera's which have failed. It is up a mast and it is pretty windy here today!
> 
> > As an aside, I seem to have picked up an additional "Forecast &" in the Forecast and Radar title bar. I cannot see where from. It is there whether I put the title in the skin or weewx.conf files...
> 
> Did you accidentally set this in skin.conf under `[Extras]radar_heading`?

No. Just been through all the locations and cannot see it. I am sure it must be where I have "hacked" some code. It will disappear when I do the next install...

J

**Reply by JonFear** [reply]
Dave

Regarding the extra "Forecast &" I have. I think I know where it is coming from but do not know how to fix it. The additional test "Forecast provided by Open Weather Maps and Live Radar provided by Windy.com" is added to the radar title in the skin.conf file for now. When this is disabled I get the standard "Forecast & Radar". My guess is that somewhere the two text boxes or whatever they are called are not lined up...

Thoughts?

Jon

**Reply by Daveiano** [reply]
The current logic is the following: 

"Forecast" is shown as heading if the user has a forecast extension enabled (by default, no way to change). If the user has a radar/webcam enabled then the "&" and "Radar" or the value `radar_heading` from skin.conf is shown.

There is currently no way to achieve a heading like "Forecast provided by Open Weather Maps and Live Radar provided by Windy.com". The best you could do would be to set `radar_heading` to "provided by Open Weather Maps and Live Radar provided by Windy.com" but then you would end up with 

"Forecast & provided by Open Weather Maps and Live Radar provided by Windy.com"

I think the current implementation is kind of a result of overthinking I guess. It would be best to let the user set the complete heading and not print static "Forecast" and "&".

I hope that makes sense now, and like I said: 

> It would be best to let the user set the complete heading and not print static "Forecast" and "&".

Any thoughts?

**Reply by JonFear** [reply]
Hi Dave

Ah As I suspected! If it is easy enough to allow the user to set the complete heading then I would say that is the way to go. I guess that updating the excellent wiki is more work than changing the code...

BW

Jon

**Reply by Daveiano** [reply]
Yeah, I think so too. 

I will release a new version within the next few days to include this change and the previous fixes. Thanks for the feedback!

**Comment by dbx-one** [comment]
Great work, keep it up. :+1:
